src/client.rs

```diff
@@ -1,398 +1,475 @@
+            recv(rx) -> msg => msg.unwrap().map(Some),
+            recv(done_rx) -> _ => Ok(None),
+            recv(ctx.timeout_rx) -> _ => Err(Error::TimedOut),
+            recv(ctx.cancel_rx) -> _ => Err(Error::Canceled),
+        };
+        self.link.unregister(key);
+        res
     }
 
     fn read_setting_inner(
-        &self,
-        group: &str,
-        name: &str,
-    ) -> Option<Result<SettingValue, Error<ReadSettingError>>> {
-        let setting = Setting::find(group, name)?;
-
-        let cgroup = match CString::new(group) {
-            Ok(cgroup) => cgroup,
-            Err(e) => return Some(Err(e.into())),
+        &mut self,
+        group: String,
+        name: String,
+        ctx: Context,
+    ) -> Result<Option<Entry>, Error> {
+        let (tx, rx) = crossbeam_channel::bounded(1);
+        let key = self.link.register({
+            let group = group.clone();
+            let name = name.clone();
+            move |msg: MsgSettingsReadResp| {
+                let fields = split_multipart(&msg.setting);
+                if fields.len() < 2 || fields[0] != group || fields[1] != name {
+                    return;
+                }
+                let _ = tx.send(Entry::try_from(msg).map(|e| {
+                    if e.value.is_some() {
+                        Some(e)
+                    } else {
+                        None
+                    }
+                }));
+            }
+        });
+        self.sender.send(MsgSettingsReadReq {
+            sender_id: Some(SENDER_ID),
+            setting: format!("{}\0{}\0", group, name).into(),
+        })?;
+        let res = crossbeam_channel::select! {
+            recv(rx) -> msg => msg.unwrap(),
+            recv(ctx.timeout_rx) -> _ => Err(Error::TimedOut),
+            recv(ctx.cancel_rx) -> _ => Err(Error::Canceled),
         };
+        self.link.unregister(key);
+        res
+    }
 
-        let cname = match CString::new(name) {
-            Ok(cname) => cname,
-            Err(e) => return Some(Err(e.into())),
+    fn write_setting_inner(
+        &mut self,
+        group: String,
+        name: String,
+        value: String,
+        ctx: Context,
+    ) -> Result<Entry, Error> {
+        let (tx, rx) = crossbeam_channel::bounded(1);
+        let key = self.link.register({
+            let group = group.clone();
+            let name = name.clone();
+            move |msg: MsgSettingsWriteResp| {
+                let fields = split_multipart(&msg.setting);
+                if fields.len() < 2 || fields[0] != group || fields[1] != name {
+                    return;
+                }
+                let _ = tx.send(Entry::try_from(msg));
+            }
+        });
+        self.sender.send(MsgSettingsWrite {
+            sender_id: Some(SENDER_ID),
+            setting: format!("{}\0{}\0{}\0", group, name, value).into(),
+        })?;
+        let res = crossbeam_channel::select! {
+            recv(rx) -> msg => msg.unwrap(),
```

Comment:
Let's replace these unwraps with descriptive expects just in case we see more reliability problems in the future.